OpenRA.Mods.RA2/Traits/CarrierParent.cs

```diff
@@ -0,0 +1,264 @@
+			}
+
+			SpawnIntoWorld(self, carrierChildEntry.Actor, self.CenterPosition);
+
+			Stack<int> spawnContainToken;
+			if (spawnContainTokens.TryGetValue(a.Info.Name, out spawnContainToken) && spawnContainToken.Any())
+				conditionManager.RevokeCondition(self, spawnContainToken.Pop());
+
+			if (loadedTokens.Any())
+				conditionManager.RevokeCondition(self, loadedTokens.Pop());
+
+			self.World.AddFrameEndTask(w =>
+			{
+				// The actor might had been trying to do something before entering the carrier.
+				// Cancel whatever it was trying to do.
+				carrierChildEntry.SpawnerChild.Stop(carrierChildEntry.Actor);
+
+				carrierChildEntry.SpawnerChild.Attack(carrierChildEntry.Actor, target);
+			});
+		}
+
+		void INotifyBecomingIdle.OnBecomingIdle(Actor self)
+		{
+			Recall(self);
+		}
+
+		void Recall(Actor self)
+		{
+			foreach (var carrierChildEntry in childEntries)
+				if (carrierChildEntry.IsLaunched && carrierChildEntry.IsValid)
+					carrierChildEntry.SpawnerChild.EnterSpawner(carrierChildEntry.Actor);
+		}
+
+		public override void OnChildKilled(Actor self, Actor child)
+		{
+			// Set clock so that regeneration happens.
+			if (respawnTicks <= 0)
+				respawnTicks = Info.RespawnTicks;
+		}
+
+		CarrierChildEntry GetLaunchable()
+		{
+			foreach (var carrierChildEntry in childEntries)
+				if (carrierChildEntry.RearmTicks <= 0 && !carrierChildEntry.IsLaunched && carrierChildEntry.IsValid)
+					return carrierChildEntry;
+
+			return null;
+		}
+
+		public IEnumerable<PipType> GetPips(Actor self)
+		{
+			if (IsTraitDisabled)
+				yield break;
+
+			var inside = 0;
+			foreach (var carrierChildEntry in childEntries)
+				if (carrierChildEntry.IsValid && !carrierChildEntry.IsLaunched)
+					inside++;
+
+			for (var i = 0; i < Info.Actors.Length; i++)
+			{
+				if (i < inside)
+					yield return info.PipType;
+				else
+					yield return PipType.Transparent;
+			}
+		}
+
+		public void PickupChild(Actor self, Actor child)
+		{
+			if (info.InstantRepair)
+			{
+				var health = child.Trait<Health>();
+				child.InflictDamage(child, new Damage(-health.MaxHP));
+			}
+
+			CarrierChildEntry childEntry = null;
+			foreach (var carrierChildEntry in childEntries)
+			{
+				if (carrierChildEntry.Actor == child)
+				{
+					childEntry = carrierChildEntry;
+					break;
+				}
+			}
+
+			if (childEntry == null)
+				throw new InvalidOperationException("An actor that isn't my child entered me?");
+
+			childEntry.IsLaunched = false;
+
+			childEntry.RearmTicks = Util.ApplyPercentageModifiers(info.RearmTicks, reloadModifiers.Select(rm => rm.GetReloadModifier()));
+
+			if (conditionManager != null && !string.IsNullOrEmpty(info.LoadedCondition))
+				loadedTokens.Push(conditionManager.GrantCondition(self, info.LoadedCondition));
+		}
+
+		public override void Replenish(Actor self, BaseSpawnerChildEntry entry)
+		{
+			base.Replenish(self, entry);
+
+			if (conditionManager != null)
+			{
+				if (!string.IsNullOrEmpty(info.LoadedCondition))
```

Comment:
This can be combined with the above check for readability:
```
if (conditionManager != null && !string.IsNullOrEmpty(info.LoadedCondition))
	loadedTokens.Push(conditionManager.GrantCondition(self, info.LoadedCondition));
```